### PR TITLE
feat: Added Shortcut functionality

### LINKED ISSUE #10: Shortcuts
**User story:**
As a user I want to be able to create file/folder shortcuts, so that I can better navigate my files

Dependent on https://github.com/HF6-PROJECT/api/issues/10

### PR DESCRIPTION
**Pull request by Anders164a**
## Description

Added shortcut functionality, to add and delete shortcut on docs, files & folders

## Type of Change

- [x] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

File: src/components/item/docs/Docs.vue [hunk 1]
```diff
@@ -5,30 +5,79 @@
...
 						class="block px-4 py-2 hover:bg-gray-100 dark:hover:bg-gray-600 dark:hover:text-white"
 						>{{ t('fileBrowser.docs.action.openInNewTab') }}</a
 					>
+					<a
+						v-else
+						:href="url(`u/docs/${modelValue.linkedItemId}`)"
+						target="_blank"
+						class="block px-4 py-2 hover:bg-gray-100 dark:hover:bg-gray-600 dark:hover:text-white"
+						>{{ t('fileBrowser.docs.action.openInNewTab') }}</a
+					>
```
Comment (kristianbinau) [inline src/components/item/docs/Docs.vue]: Do something like this instead
```suggestion
					<a
						:href="(modelValue instanceof ShortcutClass) ? url(`u/docs/${modelValue.linkedItemId}`)  : url(`u/docs/${modelValue.id}`)"
						target="_blank"
						class="block px-4 py-2 hover:bg-gray-100 dark:hover:bg-gray-600 dark:hover:text-white"
						>{{ t('fileBrowser.docs.action.openInNewTab') }}</a
					>
```

Comment (kristianbinau) [reply src/components/item/docs/Docs.vue]: In all ContextMenues where it's only 1 value that changes

File: src/components/item/docs/Docs.vue [hunk 2]
```diff
@@ -5,30 +5,79 @@
...
 						@click="editDocsModal?.open()"
 						class="block px-4 py-2 hover:bg-gray-100 dark:hover:bg-gray-600 dark:hover:text-white"
 						>{{ t('fileBrowser.docs.action.edit') }}</a
 					>
+					<a
+						v-else
+						href="javascript:void(0)"
+						@click="editShortcutModal?.open()"
+						class="block px-4 py-2 hover:bg-gray-100 dark:hover:bg-gray-600 dark:hover:text-white"
+						>{{ t('fileBrowser.shortcut.rename') }}</a
```
Comment (kristianbinau) [inline src/components/item/docs/Docs.vue]: Change all shortcut translations to match those of Folders, Files & Docs
```suggestion
						>{{ t('fileBrowser.shortcut.action.edit') }}</a
```

File: src/components/item/CreateShortcutModal.vue [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+		<hr />
+		<div class="mb-5">
+			<div
+				class="relative inline-flex w-full items-center py-1 hover:bg-gray-600"
+				@click="selectFolder(0)"
+				@mouseenter="showChildArrow(0)"
+				@mouseleave="hideChildArrow(0)"
+				id="folder0"
+			>
+				<h1 class="ml-2 mr-auto">Mine Filer</h1>
```
Comment (kristianbinau) [inline src/components/item/CreateShortcutModal.vue]: Translations????

File: src/lib/items/factory.ts [hunk 4]
```diff
@@ -1,13 +1,21 @@
...
 		if (FolderClass.isFolder(object)) return new FolderClass(object);
 		if (FileClass.isFile(object)) return new FileClass(object);
 		if (DocsClass.isDocs(object)) return new DocsClass(object);
+		if (ShortcutClass.isShortcut(object)) {
+			const shortcutClass = new ShortcutClass(object);
+
+			await shortcutClass.setLinkedItem(object.linkedItemId);
+
+			return shortcutClass;
+		}
```
Comment (kristianbinau) [inline src/lib/items/factory.ts]:
```suggestion
		if (ShortcutClass.isShortcut(object)) {
			const shortcutClass = new ShortcutClass(object);
			await shortcutClass.setLinkedItem(object.linkedItemId);
			return shortcutClass;
		}
```

File: src/lib/items/shortcuts.ts [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+	}
+
+	get linkedItem() {
+		return this._linkedItem;
+	}
+}
+
+export type ShortcutType = {
+	linkedItemId: number;
+	linkedItem: FileClass | ShortcutClass | undefined;
```
Comment (kristianbinau) [inline src/lib/items/shortcuts.ts]:
```suggestion
	linkedItem: FileClass | ShortcutClass | DocsType | undefined;
```

File: src/lib/items/shortcuts.ts [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+		}
+
+		if (json.mimeType === 'application/vnd.cloudstore.docs') {
+			this._linkedItem = new DocsClass(json);
+			return;
+		}
+
+		if ('blobUrl' in json && typeof json.blobUrl === 'string') {
+			this._linkedItem = new FileClass(json);
+		}
```
Comment (kristianbinau) [inline src/lib/items/shortcuts.ts]: Use `Folder.isFolder()` and the others

File: src/lib/items/shortcuts.ts [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+				throw new Error(json.error);
+			}
+
+			throw new Error(await response.text());
+		}
+	}
+
+	static isShortcut(object: any): object is ShortcutType {
+		if (!ItemClass.isItem(object)) return false;
+
```
Comment (kristianbinau) [inline src/lib/items/shortcuts.ts:118]: Maybe also check if object has `linkedItemId` and that it is a number

File: src/lib/items/shortcuts.ts [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+import { FolderClass } from './folders';
+import { FileClass } from './files';
+import { DocsClass } from './docs';
+
+export class ShortcutClass extends ItemClass {
+	public _linkedItemId: number;
+	public _linkedItem: FolderClass | FileClass | DocsClass | undefined;
+	public color: string | undefined;
+	public text: string | undefined;
+	public blobUrl: string | undefined;
```
Comment (kristianbinau) [inline src/lib/items/shortcuts.ts]: Wat

**Review by kristianbinau — CHANGES_REQUESTED** [review]

**Review by frederikpyt — DISMISSED** [review]

**Review by frederikpyt — APPROVED** [review]

**Review by kristianbinau — APPROVED** [review]

**Pull request merged**